### PR TITLE
[Hub Generated] Review request for Microsoft.Maps/StyleRecipe to add version preview/2022-01-01-preview, preview/2022-01-01-preview and preview/2022-01-01-preview

### LINKED ISSUE #17312: [Maps - Styling] REST API Review for Azure Maps Styling Service
Service Name: Azure Maps Styling Service

Key contact for the review: Xubin Zhuge(@Xubin Zhuge) / James Jonghan Park(@James Jonghan Park)

The changes contain both existing APIs(Tileset) and new APIs(StyleRecipe, StyleSet)

The service is deployed, but the proposed changes are still in progress.

What changed in this PR:
 - We create a new 2022-01-01-preview version API for Tileset, StyleSet and StyleRecipe under Creator
- We extended our exising Tileset API with Tileset Recipe ingested, and we also added Tileset Recipe GET and Validation API.
- We added two new set of APIs StyleSet and StyleRecipe which are our new APIs

Custom Styling Design doc: 

https://microsoft.sharepoint.com/:w:/t/MicrosoftMapsGeospatialTeam/EUyAGqR7R4pJoIXkXQdlQcAB7SPLJQImjNS0eQGp8xathQ?e=eccr4D

### PR DESCRIPTION
**Pull request by zhugexubin**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/zhugexubin/azure-rest-api-specs/dev-maps-Microsoft.Maps-xubin-2022-01-01-preview...main/).

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [x] new service onboarding
      - [x] new API version
      - [x] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [x] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/maps/data-plane/Microsoft.Maps/Tileset/preview/2022-01-01-preview/tileset.json [hunk 1]
```diff
@@ -594,8 +711,100 @@
...
+      "required": [
+        "featureClasses"
+      ],
+      "properties": {
+        "featureClasses": {
+          "description": "The list of featureClasses that will be used to build the layer.",
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
```
Comment (tjprescott) [inline specification/maps/data-plane/Microsoft.Maps/Tileset/preview/2022-01-01-preview/tileset.json:770]: What kinds of values are permissible here? Should this be an enum? How is a customer supposed to know what they can supply?

File: specification/maps/data-plane/Microsoft.Maps/Tileset/preview/2022-01-01-preview/tileset.json [hunk 2]
```diff
@@ -594,8 +711,100 @@
...
+        },
+        "filter": {
+          "description": "Removes features that do not match a pattern.",
+          "type": "object"
```
Comment (tjprescott) [inline specification/maps/data-plane/Microsoft.Maps/Tileset/preview/2022-01-01-preview/tileset.json:774]: This is a very vague type. Is this a dictionary?

File: specification/maps/data-plane/Microsoft.Maps/Tileset/preview/2022-01-01-preview/tileset.json [hunk 3]
```diff
@@ -594,8 +711,100 @@
...
+          "items": {
+            "type": "string"
+          }
+        },
+        "errors": {
+          "description": "The list of validation errors",
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
```
Comment (tjprescott) [inline specification/maps/data-plane/Microsoft.Maps/Tileset/preview/2022-01-01-preview/tileset.json:798]: So even if there are errors, the expected response would be 200 for a call to validate?

File: specification/maps/data-plane/Microsoft.Maps/Tileset/preview/2022-01-01-preview/tileset.json (hunk 1 above)
Comment (zhugexubin) [reply specification/maps/data-plane/Microsoft.Maps/Tileset/preview/2022-01-01-preview/tileset.json:770]: FeatureClasses are defined in our Ontology [ontology](https://docs.microsoft.com/en-us/azure/azure-maps/creator-facility-ontology?pivots=facility-ontology-v2), The values here are mainly generated from customer data which is not necessarily a static set of values. Customer should know their data well before they can provide any value here and we also have validation logic in the backend to double check the values.

File: specification/maps/data-plane/Microsoft.Maps/Tileset/preview/2022-01-01-preview/tileset.json (hunk 2 above)
Comment (zhugexubin) [reply specification/maps/data-plane/Microsoft.Maps/Tileset/preview/2022-01-01-preview/tileset.json]: This is actually a Json array which represent a mapbox expression, the element of the array can be another expression. So I changed it to be an array of object instead.

File: specification/maps/data-plane/Microsoft.Maps/Tileset/preview/2022-01-01-preview/tileset.json (hunk 3 above)
Comment (zhugexubin) [reply specification/maps/data-plane/Microsoft.Maps/Tileset/preview/2022-01-01-preview/tileset.json:801]: Yes, the valid response will always be 200 with response body containing the validation details for the tileset recipe request body. The response can be 400 BadRequest when customer gives an invalid input, for example, the datasetId parameter provided does not exist.

**Comment by JeffreyRichter** [comment]
Notes:
- Remove non-2xx status codes from swagger; we only put success codes in the swagger
- Make sure the default http error refers to the command Azure error response
- We prefer PUT instead of POST for creating a new resource
- We have a greatly improved pattern for LROs; please use it: https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#long-running-operations--jobs

**Pull request closed without merging**